DoLah/Parser/DateTimeParser.cpp

```diff
@@ -79,14 +98,171 @@ namespace DoLah {
         return std::stoi(year);
     }
 
+    int DateTimeParser::getDate(std::string str) {
+        std::string out;
+        for (size_t d = 0; d < datePattern.size(); d++) {
+            if (std::regex_match(str, std::regex(datePattern.at(d), std::regex_constants::icase))) {
+                return (int)d;
+            }
+        }
 
+        return REJECT;
+    }
+
+    int DateTimeParser::getTime(std::string str) {
+        str = ParserLibrary::tolowercase(str);
+
+        int time = 0;
+        bool isTime = false;
+
+        bool isPM = false;
+        bool isAM = false;
+        if (str.find(PM) != std::string::npos) {
+            ParserLibrary::stringRemove(str, PM);
+            isPM = true;
+            isTime = true;
+        } else if (str.find(AM) != std::string::npos) {
+            ParserLibrary::stringRemove(str, AM);
+            isAM = true;
+            isTime = true;
+        }
+        
+        std::vector<std::string> strArr = ParserLibrary::explode(str, timeDivider);
+
+        int hour = std::stoi(strArr.at(0));
+        time = hour * 60;
+
+        if (isAM && hour >= 12) {
+            throw std::invalid_argument("");
+        }
+
+        if (isPM && hour < 12) {
+            time += 12 * 60;
+        }
+
+        if (strArr.size() > 1) {
+            time += std::stoi(strArr.at(1));
+            isTime = true;
+        }
+
+        if (!isTime) {
+            throw std::invalid_argument("");
+        }
+
+        return time;
+    }
+
+    int DateTimeParser::getDateModifier(int date, bool notThisWeek) {
+        time_t t = time(0);
+        std::tm current;
+        localtime_s(&current, &t);
+
+        int diff = date - current.tm_wday + 1;
+        if (diff < 0) {
+            diff = 7 + diff;
+        } else if (notThisWeek) {
+            diff += 7;
+        }
+
+        return diff;
+    }
 
-    std::tm DateTimeParser::toDateFormat(std::vector<std::string> strArr) {
+    std::tm DateTimeParser::checkRelativeDateFormat(std::vector<std::string> strArr) {
         std::tm output;
+        int dayDiff = 0;
+        int weekDiff = 0;
+        int monthDiff = 0;
+
+        int size = strArr.size();
+
+        int index = 0;
+        std::string element;
+        
+        element = strArr.at(index++);
+        int date = getDate(element);
+        if (strArr.size() == 1) { // singleton format
```

Comment:
Literally singleton format. Check the meaning of the English word.

It basically means there is a single part.